src/tir/transforms/unify_thread_binding.cc

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file unify_thread_binding.cc
+ */
+
+#include <tvm/tir/analysis.h>
+#include <tvm/tir/stmt_functor.h>
+#include <tvm/tir/transform.h>
+
+#include "ir_utils.h"
+
+namespace tvm {
+namespace tir {
+
+/*!
+ * \brief A mutator which searches AttrStmts of thread bindings and changes the `node` field IterVar
+ * of the AttrStmts, so that for one kind of thread binding (except "vthread"), all such thread
+ * bindings use the same IterVar
+ */
+class ThreadBindingUnifier : public StmtExprMutator {
+ public:
+  static Stmt Unify(const PrimFunc& f) { return ThreadBindingUnifier().VisitStmt(f->body); }
+
+ private:
+  Stmt VisitStmt_(const AttrStmtNode* attr) final {
+    // If this AttrStmt is not thread binding attribute, return as usual.
+    if (attr->attr_key != attr::thread_extent && attr->attr_key != attr::virtual_thread) {
+      return StmtMutator::VisitStmt_(attr);
+    }
+
+    // Step 1. Fetch the old IterVar.
+    IterVar old_iter_var = Downcast<IterVar>(attr->node);
+    IterVar new_iter_var;
+
+    // Step 2. See if an IterVar for this kind of thread binding was created before. If so, we use
+    // the created IterVar. Otherwise, we create a new IterVar for this thread binding and store the
+    // IterVar in mapping `thread_tag2iter_var_map_`.
+    Map<String, IterVar>::iterator it = thread_tag2iter_var_map_.find(old_iter_var->thread_tag);
+    if (it != thread_tag2iter_var_map_.end()) {
+      new_iter_var = (*it).second;
+      CHECK(ExprDeepEqual()(old_iter_var->dom->extent, (*it).second->dom->extent))
```

Comment:
Please use `analyzer::CanProveEqual` instead of ExprDeepEqual

Reply:
Okay. But I'm curious about the difference between them :thinking:.